jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,735 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.pathmap.PathSpec;
+import org.eclipse.jetty.http.pathmap.PathSpecSet;
+import org.eclipse.jetty.server.ConnectionMetaData;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IncludeExclude;
+import org.eclipse.jetty.util.IncludeExcludeSet;
+import org.eclipse.jetty.util.InetAddressPattern;
+import org.eclipse.jetty.util.TypeUtil;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An abstract {@link Handler.Wrapper} that may conditionally handle a {@link Request}.
+ * The conditions are implemented by {@link IncludeExclude}s of:
+ * <ul>
+ *     <li>A method name, which can be efficiently matched</li>
+ *     <li>A {@link PathSpec} or sting representation, which can be efficient matched.</li>
+ *     <li>An arbitrary {@link Predicate} taking the {@link Request}, which is matched in a linear test of all predicates.</li>
+ * </ul>
+ * <p>If the conditions are met, the abstract {@link #doHandle(Request, Response, Callback)} method will be invoked,
+ * otherwise the {@link #doNotHandle(Request, Response, Callback)} method will be invoked, which by default calls
+ * the {@link #nextHandle(Request, Response, Callback)} method.</p>
+ * <p>A typical usage is to extend the class and provide an implementation of {@link #doHandle(Request, Response, Callback)}:</p>
+ * <pre>{@code
+ * public class MyOptionalHandler extends ConditionalHandler
```

Comment:
Javadoc is now outdated and needs to be revisited.

Do we want users to extend directly `ConditionalHandler`?
If not, then it should be a `sealed` class.

Reply:
The javadoc is not out of date and has already been updated.

We want user to directly extend `ConditionalHandler`.  That is is the case in which they provide a `doHandle` method.
There are also some concrete implementations with specific `doHandle` implementations.